### ISSUE TITLE
Investigate and optimize solr timeouts

### ISSUE THREAD

**Issue by jrochkind** [sciencehistory/scihist_digicoll#2684]
We *intend* to set a timeout on our app's connection to solr, to keep it from waiting indefinitely on a non-responsive solr.  Becuase that happens sometime, and we don't want to tie up all our web workers waiting on non-responsive solr. 

We intend to have it wait only 4 seconds for solr before giving up:

https://github.com/sciencehistory/scihist_digicoll/blob/53a1f0cd346f7456f547f37d9c5cf0852a419a1b/config/blacklight.yml#L10

Also, it is intended to retry twice, waiting only a fraction of a second between retries. 

https://github.com/sciencehistory/scihist_digicoll/blob/53a1f0cd346f7456f547f37d9c5cf0852a419a1b/app/models/scihist/blacklight_solr_repository.rb#L40-L45

This means that if a Solr is being consistently unresponsive (or taking 10 seconds, or jsut more than 4, to respond), the app ought to wait:  4 seconds + ~1 second delay + 4 seconds + ~1 second delay + 4 seconds before giving up. Or around 14 seconds. 

Which may already be actually far too long to tie up a worker on unresponsive solr? Maybe we want only ONE retry. (we have a retry in because in some cases SearchStax solr was being weird and a retry worked).  Maybe less than 4 second timeout?  (Although solr can be unpredictably slow). 

But additionally, it's not clear to me that it really _is_ working like this. 

In times of Solr unresponsiveness, we are having heroku tell us lot sof responses are timing out at _30 seconds_, heroku's own timeout. Suggesting they were being held up waiting for solr in sum mucb longer than the 14 seconds we expect too. 

So it's possible the configuration isn't working like we want anyway -- it was a bit convoluted to try to set it up. 

**investigate how solr timeout/retry is _actually_ working, with _actual_ max time spent waiting for solr, then consider if we want to change it to keep app more reliable even when solr is not**

**Comment by eddierubeiz** [comment]
**Relevant lit**
 - (Various stackoverflow articles)
 - https://devcenter.heroku.com/articles/request-timeout
 - https://devcenter.heroku.com/articles/h12-request-timeout-in-ruby-mri
 - https://devcenter.heroku.com/articles/error-codes#h12-request-timeout
 - https://scoutapm.com/blog/response-time-analysis

**Tools**
 - HoneyBadger's [list of recent timeout errors](https://app.honeybadger.io/projects/58989/faults?q=TimeoutError);
    - For each of these errors, you can look at each occurrence's `request_id`, which you can then search for in the Heroku logs. Take request [835e5b15-9fb1-449c-aa45-bb26d3d8e5bf](https://my.na-01.cloud.solarwinds.com/218426623640663040/logs?q=835e5b15-9fb1-449c-aa45-bb26d3d8e5bf), which I believe to be typical, as an example.
				Note the two four-second timeouts (12:01:46 - 12:01:50, then 12:01:50 - 12:01:54), with the server returning an error to the user immediately afterwards.
- The Heroku logs, in which a search for the phrase `code=H12` gives a list of recent genuine 30-second Heroku timeouts. Our `H12` errors are coming not from SearchStax, but from two other sources:
    - https://digital.sciencehistory.org/admin/fixity_report, which has gotten really slow (out of scope, but we can certainly make it better if we want);
   - the FAST subject-heading lookup service (which is never fast and often down).
 - Scout's [external services](https://scoutapm.com/apps/347458/external_services) report.
		

**Findings**
 - Our Blacklight retry code (listed above) is working exactly as it should, as far as I can tell.
 - During November 2024 at least, the only H12 timeouts we've had affected staff users.
 - I have yet to see an H12 timeout that we can attribute to SearchStax, and I've looked at a bunch of Faraday timeouts.

**Suggestions**
Let's consider leaving the retry code as is, but changing the timeout from 4 to just 2 seconds.

**Comment by jrochkind** [comment]
Thanks for looking into this!

I could SWEAR I saw timeouts related to solr... but I"m again wondering if it was *writing* to solr.  At any rate, if we can't find now, so be it. I guess if I find one again, I'll re-open, and document better!

Solr _can_ be slow sometimes... I'm not sure how much more often we'd be giving people errors if we reduced to 2 seconds?  (We DO have one retry configured for solr errors though, I think...)

**Comment by eddierubeiz** [comment]
We have a [ticket to add a timeout to Fast](https://github.com/sciencehistory/scihist_digicoll/issues/2770) already.

**Comment by eddierubeiz** [comment]
Creating a ticket for slow fixity report now...

**Comment by eddierubeiz** [comment]
I just created [an alert in HB](https://my.na-01.cloud.solarwinds.com/218426623640663040/alerts/ae195758-bcc4-4396-8196-87d6f5f3d448/overview?duration=3600) which will email us next time we get an H12 for any reason.
Maximum one email a day; if it becomes annoying we can turn it off.

**Comment by eddierubeiz** [comment]
We got two good examples of H12 timeouts caused by SOLR. Log extracts for both requests below, with some redactions for readability.

**Comment by eddierubeiz** [comment]
`01:57:28 PM  app/web.1 W, [13:57:28.640695 #12137]  WARN -- : [FIRST_ID] Scihist::BlacklightSolrRepository: Retrying Solr request: HTTP 200: Faraday::TimeoutError: retry 1: will try again in 0.3s`

**Time interval: 4.3 seconds**

```
01:57:33 PM  app/web.1 W, [13:57:32.952093 #12137]  WARN -- : [FIRST_ID] Scihist::BlacklightSolrRepository: Retrying Solr request: HTTP 200: Faraday::TimeoutError: retry 2: will try again in 0.6s
```

**Time interval: 6.7 seconds**

```
01:57:39 PM  heroku/router at=error code=H12 desc="Request timeout" method=GET path="/catalog?f%255Bcreator_facet%255D%255B%255D=Kwolek%2C%2BStephanie%2C%2B1923-2014&page=36&range%5Byear_facet_isim%5D%5Bbegin%5D=1980&range%5Byear_facet_isim%5D%5Bend%5D=1999&sort=oldest_date" host=digital.sciencehistory.org request_id=FIRST_ID fwd="51.222.253.3" dyno=web.1 connect=0ms service=30000ms status=503 bytes=0 protocol=https
```

**Time interval: 13.8 seconds**

```
01:57:53 PM  app/web.1 I, [13:57:52.765211 #12137]  INFO -- : [FIRST_ID] method=GET path=/catalog?f%255Bcreator_facet%255D%255B%255D=Kwolek%2C%2BStephanie%2C%2B1923-2014&page=36&range%5Byear_facet_isim%5D%5Bbegin%5D=1980&range%5Byear_facet_isim%5D%5Bend%5D=1999&sort=oldest_date format=*/* controller=CatalogController action=index status=500 allocations=385311 duration=43708.69 view=0.00 db=6.92 ua=bot:aHrefs_Bot ip=51.222.253.3
01:57:53 PM  app/web.1 E, [13:57:52.767059 #12137] ERROR -- : [FIRST_ID]
01:57:53 PM  app/web.1 [FIRST_ID] Faraday::TimeoutError (Net::ReadTimeout with #<TCPSocket:(closed)>):
01:57:53 PM  app/web.1 [FIRST_ID]
01:57:53 PM  app/web.1 [FIRST_ID] Net::ReadTimeout (Net::ReadTimeout with #<TCPSocket:(closed)>)
01:57:53 PM  app/web.1 [FIRST_ID] Net::ReadTimeout (Net::ReadTimeout with #<TCPSocket:(closed)>)
01:57:53 PM  app/web.1 [FIRST_ID]
01:57:53 PM  app/web.1 I, [13:57:52.771037 #12137]  INFO -- : [FIRST_ID] method=GET path=/500?f%255Bcreator_facet%255D%255B%255D=Kwolek%2C%2BStephanie%2C%2B1923-2014&page=36&range%5Byear_facet_isim%5D%5Bbegin%5D=1980&range%5Byear_facet_isim%5D%5Bend%5D=1999&sort=oldest_date format=*/* controller=ErrorsController action=internal_error status=500 allocations=909 duration=2.03 view=1.61 db=0.00 ua=bot:aHrefs_Bot ip=51.222.253.3
```

**Comment by eddierubeiz** [comment]
```
01:57:28 PM  app/web.1 W, [13:57:28.816274 #12101]  WARN -- : [SECOND_ID] Scihist::BlacklightSolrRepository: Retrying Solr request: HTTP 200: Faraday::TimeoutError: retry 1: will try again in 0.3s
```

**Time interval: 4.3 seconds**

```
01:57:33 PM  app/web.1 W, [13:57:33.126676 #12101]  WARN -- : [SECOND_ID] Scihist::BlacklightSolrRepository: Retrying Solr request: HTTP 200: Faraday::TimeoutError: retry 2: will try again in 0.6s
```

**Time interval: 10.4 seconds**

```
01:57:43 PM  heroku/router at=error code=H12 desc="Request timeout" method=GET path="/collections/gt54kn818?ID=126&Numb=1&range%5Byear_facet_isim%5D%5Bbegin%5D=1995&range%5Byear_facet_isim%5D%5Bend%5D=1995&sort=newest_date" host=digital.sciencehistory.org request_id=SECOND_ID fwd="51.222.253.3" dyno=web.1 connect=0ms service=30000ms status=503 bytes=0 protocol=https
```

**Time interval: 9.3 seconds**

```
01:57:52 PM  app/web.1 I, [13:57:52.476808 #12101]  INFO -- : [SECOND_ID] method=GET path=/collections/gt54kn818?ID=126&Numb=1&range%5Byear_facet_isim%5D%5Bbegin%5D=1995&range%5Byear_facet_isim%5D%5Bend%5D=1995&sort=newest_date format=*/* controller=CollectionShowControllers::OralHistoryCollectionController action=index status=500 allocations=138971 duration=38934.86 view=0.00 db=8.39 ua=bot:aHrefs_Bot ip=51.222.253.3
01:57:52 PM  app/web.1 E, [13:57:52.478770 #12101] ERROR -- : [SECOND_ID]
01:57:52 PM  app/web.1 [SECOND_ID] Faraday::TimeoutError (Net::ReadTimeout with #<TCPSocket:(closed)>):
01:57:52 PM  app/web.1 [SECOND_ID]
01:57:52 PM  app/web.1 [SECOND_ID] Net::ReadTimeout (Net::ReadTimeout with #<TCPSocket:(closed)>)
01:57:52 PM  app/web.1 [SECOND_ID] Net::ReadTimeout (Net::ReadTimeout with #<TCPSocket:(closed)>)
01:57:52 PM  app/web.1 [SECOND_ID]
01:57:52 PM  app/web.1 I, [13:57:52.481067 #12101]  INFO -- : [SECOND_ID] method=GET path=/500?ID=126&Numb=1&range%5Byear_facet_isim%5D%5Bbegin%5D=1995&range%5Byear_facet_isim%5D%5Bend%5D=1995&sort=newest_date format=*/* controller=ErrorsController action=internal_error status=500 allocations=602 duration=0.83 view=0.51 db=0.00 ua=bot:aHrefs_Bot ip=51.222.253.3
```

**Comment by eddierubeiz** [comment]
First request took up 24.8 seconds
Second request took 24.0 seconds.

In both cases, the H12 was logged well before the Faraday timeout error.

Both requests were from the same IP address, `51.222.253.3`, from user agent `Hrefs_Bot.`

**Comment by eddierubeiz** [comment]
As detailed above, I believe `Faraday::TimeoutError` and `Net::ReadTimeout` errors are happening as they should, and _mostly_ preventing H12 errors (in which the Heroku router gives up on serving the request after 30 seconds because the dynos have not yet delivered a response.).

I spent some more time investigating this after we experienced two H12 timeouts associated with slow SOLR responses on Nov. 24th. I haven't been able to replicate these, so the connection between the `net` and `faraday` timeouts and the H12 timeouts is unclear to me. (When I subjected our staging app to several near-simultaneous calls to a mock SOLR server that I deliberately slowed down, `BlacklightSolrRepository` did what it was supposed to do, and stopped trying to contact the SOLR server after two unsuccesful retries.)

The good news is that these solr-associated H12 timeouts appear to be rare: they happen roughly once a month at the most, but the mystery remains.
